src/libostree/ostree-sysroot-deploy.c

```diff
@@ -1073,40 +1124,79 @@ swap_bootlinks (OstreeSysroot *self,
+                                                      ostree_deployment_get_bootcsum (deployment),
+                                                      NULL);
+      g_autofree char *bootlink_pathname = g_strdup_printf ("%s/%d", bootlink_parent, ostree_deployment_get_bootserial (deployment));
       g_autofree char *bootlink_target = g_strdup_printf ("../../../deploy/%s/deploy/%s.%d",
-                                                       ostree_deployment_get_osname (deployment),
-                                                       ostree_deployment_get_csum (deployment),
-                                                       ostree_deployment_get_deployserial (deployment));
-      g_autoptr(GFile) linkname = g_file_get_child (ostree_subbootdir, bootlink_pathname);
-      g_autoptr(GFile) linkname_parent = g_file_get_parent (linkname);
+                                                          ostree_deployment_get_osname (deployment),
+                                                          ostree_deployment_get_csum (deployment),
+                                                          ostree_deployment_get_deployserial (deployment));
 
-      if (!ot_util_ensure_directory_and_fsync (linkname_parent, cancellable, error))
+      if (!glnx_shutil_mkdir_p_at (ostree_subbootdir_dfd, bootlink_parent, 0755, cancellable, error))
         goto out;
-      
-      if (!g_file_make_symbolic_link (linkname, bootlink_target, cancellable, error))
+
+      if (!symlink_at_replace (bootlink_target, ostree_subbootdir_dfd, bootlink_pathname,
+                               cancellable, error))
         goto out;
     }
 
-  if (!ot_gfile_atomic_symlink_swap (ostree_bootdir, ostree_subbootdir_name,
-                                     cancellable, error))
+  ret = TRUE;
+ out:
+  return ret;
+}
+
+static gboolean
+swap_bootlinks (OstreeSysroot *self,
+                int            bootversion,
+                GPtrArray     *new_deployments,
+                GCancellable  *cancellable,
+                GError       **error)
+{
+  gboolean ret = FALSE;
+  int old_subbootversion, new_subbootversion;
+  glnx_fd_close int ostree_dfd = -1;
+  glnx_fd_close int ostree_subbootdir_dfd = -1;
+  g_autofree char *ostree_bootdir_name = NULL;
+  g_autofree char *ostree_subbootdir_name = NULL;
+
+  if (!glnx_opendirat (self->sysroot_fd, "ostree", TRUE, &ostree_dfd, error))
     goto out;
 
+  ostree_bootdir_name = g_strdup_printf ("boot.%d", bootversion);
+
+  if (bootversion != self->bootversion)
+    {
+      if (!_ostree_sysroot_read_current_subbootversion (self, bootversion, &old_subbootversion,
+                                                        cancellable, error))
+        goto out;
+    }
+  else
+    old_subbootversion = self->subbootversion;
+
+  new_subbootversion = old_subbootversion == 0 ? 1 : 0;
+
+  ostree_subbootdir_name = g_strdup_printf ("boot.%d.%d", bootversion, new_subbootversion);
```

Comment:
I wish the previous 17 lines weren't duplicated from `create_new_bootlinks` above